### PR TITLE
Implement js_map bindings

### PR DESCRIPTION
**Pull request by purefunctor**
This adds the initial bindings for the [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) object. A couple of questions:

* Do you think I should add an interface file for the documentation?
* If so, do you reckon I can do it in a separate PR so I can add it for `js_set` as well?

### CODE REVIEW

**Comment by purefunctor** [comment]
Not sure what broke in the Opam pipeline but it seems to be marking a cram test to be dirty. https://github.com/melange-re/melange/actions/runs/8595735244/job/23550961309?pr=1101#step:6:885

**Comment by anmonteiro** [comment]
> This adds the initial bindings for the [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) object.

I don't know how we missed that recently. Thanks for catching this.

> Not sure what broke in the Opam pipeline

yeah this is because of the new dune release. I'll fix the tests in a separate PR.

**Comment by anmonteiro** [comment]
tests fixed in https://github.com/melange-re/melange/pull/1102, CI should be green if you rebase.

**Comment by purefunctor** [comment]
Tests all green!

**Comment by anmonteiro** [comment]
> * Do you think I should add an interface file for the documentation?

We haven't added them because they're most `external` declarations which would just be duplicated. If the generated documentation doesn't show the docstrings for these modules, then we should probably consider adding interface files. Can you check?

File: jscomp/runtime/js_map.ml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-(* Copyright (C) 2022- Authors of Melange
+(* Copyright (C) 2024- Authors of Melange
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:1]:
```suggestion
(* Copyright (C) 2022- Authors of Melange
```

File: jscomp/runtime/js_map.ml [hunk 2]
```diff
@@ -24,4 +24,31 @@
 
 (** ES6 Map API *)
 
+open Melange_mini_stdlib
+
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:28]:
```suggestion
```

File: jscomp/runtime/js_map.ml [hunk 3]
```diff
@@ -24,4 +24,31 @@
...
 type ('k, 'v) t
+
+external make : unit -> ('k, 'v) t = "Map" [@@mel.new]
+external fromArray : ('k * 'v) array -> ('k, 'v) t = "Map" [@@mel.new]
+external toArray : ('k, 'v) t -> ('k * 'v) array = "Array.from"
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:33]:
```suggestion
external toArray : ('k, 'v) t -> ('k * 'v) array = "from" [@@mel.scope "Array"]
```

File: jscomp/runtime/js_map.ml [hunk 4]
```diff
@@ -24,4 +24,31 @@
...
+external size : ('k, 'v) t -> int = "size" [@@mel.get]
+external has : key:'k -> bool = "has" [@@mel.send.pipe: ('k, 'v) t]
+external unsafeGet : key:'k -> 'v = "get" [@@mel.send.pipe: ('k, 'v) t]
+
+let get ~key map =
+  if map |> has ~key then Some (unsafeGet ~key map) else None
+
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:40]:
```suggestion
```

File: jscomp/runtime/js_map.ml [hunk 5]
```diff
@@ -24,4 +24,31 @@
...
+external set : key:'k -> value:'v -> ('k, 'v) t = "set"
+[@@mel.send.pipe: ('k, 'v) t]
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:42]:
```suggestion
external set : key:'k -> value:'v -> ('k, 'v) t = "set" [@@mel.send]
```

File: jscomp/runtime/js_map.ml [hunk 6]
```diff
@@ -24,4 +24,31 @@
...
+
+external clear : ('k, 'v) t -> unit = "clear" [@@mel.send]
+external delete : key:'k -> bool = "delete" [@@mel.send.pipe: ('k, 'v) t]
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:45]:
```suggestion
external delete : ('k, 'v) t -> key:'k -> bool = "delete" [@@mel.send]
```

File: jscomp/runtime/js_map.ml [hunk 7]
```diff
@@ -24,4 +24,31 @@
...
+
+external forEach : f:(('v -> 'k -> ('k, 'v) t -> unit)[@u]) -> unit = "forEach"
+[@@mel.send.pipe: ('k, 'v) t]
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:48]:
```suggestion
external forEach : f:(('v -> 'k -> ('k, 'v) t -> unit)[@mel.uncurry]) -> unit = "forEach"
[@@mel.send.pipe: ('k, 'v) t]
```

File: jscomp/runtime/js_map.ml [hunk 8]
```diff
@@ -24,4 +24,31 @@
...
+
+external keys : 'k Js.iterator = "keys" [@@mel.send.pipe: ('k, 'v) t]
+external values : 'v Js.iterator = "values" [@@mel.send.pipe: ('k, 'v) t]
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:51]:
```suggestion
external keys : ('k, 'v) t -> 'k Js.iterator = "keys" [@@mel.send]
external values : ('k, 'v) t -> 'v Js.iterator = "values" [@@mel.send]
```

File: jscomp/runtime/js_map.ml [hunk 9]
```diff
@@ -24,4 +24,31 @@
...
+
+external entries : ('k * 'v) Js.iterator = "entries"
+[@@mel.send.pipe: ('k, 'v) t]
```
Comment (anmonteiro) [inline jscomp/runtime/js_map.ml:54]:
```suggestion
external entries : ('k, 'v) t -> ('k * 'v) Js.iterator = "entries" [@@mel.send]
```

**Comment by anmonteiro** [comment]
thank you!

**Pull request merged**